### PR TITLE
Adding smoothsort

### PR DESCRIPTION
**Pull request by kanishkgits**
## Description

Smoothsort is a sorting algorithm that efficiently handles partially sorted data, using specialized heaps (Leonardo heaps) to optimize performance. It’s in-place, with a best-case time complexity of O(n).
Some helpful articles:
Wikipedia: https://en.wikipedia.org/wiki/Smoothsort
GeeksforGeeks: https://www.geeksforgeeks.org/introduction-to-smooth-sort

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I ran below commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

File: src/sorting/smooth_sort.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
+//smooth sort
+
+
+pub fn smooth_sort(nums: &mut Vec<i32>) {
+    let n = nums.len();
```
Comment (vil02) [inline src/sorting/smooth_sort.rs:5]: Why not to use `nums.len()` directly?

File: src/sorting/smooth_sort.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            current = right_child;
+        } else {
+            break;
+        }
+
+        heap_size -= 1;
+    }
+}
+
+//tests
```
Comment (vil02) [inline src/sorting/smooth_sort.rs:67]:
```suggestion
```

**Review by vil02 — CHANGES_REQUESTED** [review]

**Comment by sozelfist** [comment]
Please solve all the issues related to fmt, and clippy and make sure all the tests passed.

**Comment by kanishkgits** [comment]
It's a complex algorithm and I was unable to pass the cargo test :(

**Pull request closed without merging**